## This utility script converts the phylogeny and the P/A matrix into the format expected by Count.jar and generates the annotation table expected by Count using the eggNOG data.
##
## USAGE
## Rscript convert_ASR_inputs.R @tree @pa @ann @out_dir
##
## PARAMETERS
## tree     a phylogeny in Newick format, generated by PhyloPhlAn, probably with mixed taxa and number node labels
## pa       presence/absence pangenome matrix in csv format exported by PPanGOLiN
## ann      eggnog-mapper annotation table in csv format
## out_dir  output directory for the converted inputs

library(ape)
library(dplyr)
library(data.table)
library(stringr)

args = commandArgs(trailingOnly = TRUE)
tree.file = args[1]
pa.file = args[2]
ann.file = args[3]
out.dir = args[4]

## Converting the tree
# Mixed taxa and numbered node labels are fully converted to node labels to more easily track a node in the Count output
tree = read.tree(tree.file)
tree$node.label = as.character(1:tree$Nnode)
write.tree(tree, file = paste(out.dir, "input_ready.tree", sep = "/"))

## Converting the P/A data
# Gene family presences are converted into a plain count number instead of lists of genomic intervals
pa = as.data.frame(fread(pa.file, drop = 2:14, sep = " "))
rownames(pa) = pa$Gene
pa[] = lapply(pa, function (x) str_count(x, ",") + 1) # Genomic intervals are separated by commas
pa$Gene = rownames(pa)
new.cols = colnames(pa)
new.cols = lapply(new.cols, function (x) str_replace_all(x, '_', ' ')) # Remove underscores as Count ignores these
colnames(pa) = new.cols
fwrite(pa, file = paste(out.dir, "matrix_counts.tsv", sep = "/"), sep = "\t")

## Preparing the eggNOG annotation table
cols.pa = c(1,3)
cols.ann = c(1,7)
pa = as.data.frame(fread(pa.file, select = cols.pa, sep = ","))
ann.filt = system(paste("cat", ann.file, "| grep -v '##'"), intern = TRUE)
ann = as.data.frame(fread(text = ann.filt, select = cols.ann, sep = "\t"))
ann = rename(ann, 'Gene' = '#query')
joined_ann = inner_join(pa, ann)
fwrite(joined_ann, file = paste(out.dir, "matrix_annotations.tsv", sep = "/"), sep = "\t", quote = FALSE, col.names = FALSE)
